### ISSUE TITLE
Don't force uniqueness in target names within a project

### ISSUE THREAD

**Issue by nirbheek** [mesonbuild/meson#2861]
Currently, we don't allow you to have two `executable()` or `shared_library()` targets of the same name in the same project. However, this is sometimes needed:

1. [Test and benchmark executables can have the same names](https://github.com/mesonbuild/meson/pull/1867#issuecomment-304692828)
2. [A library and its Python module can have the same name](https://github.com/mesonbuild/meson/issues/2859#issuecomment-354684303)

This restriction seems to have been borrowed from cmake, but there is nothing in our code stopping us from implementing it, as shown by [this PR](https://github.com/mesonbuild/meson/pull/1867), and no other build system has this restriction.

In practice, having two targets of the same type with the same name, and in different subdirs (or if in the same subdir, with different suffixes and/or prefixes) should not be a problem for the build system (unless they are installed into the same destdir or you're building with a flat layout), and is anyway equivalent (in practice) to having a target with the same name in a subproject (which we don't, shouldn't, and can't really disallow).

We should remove this restriction.

CCing @kaimast

**Comment by NickeZ** [comment]
Having the same name of a library and an executable will fail if you build the project with `msbuild` on windows. I think uniqueness should be enforced even stricter than it currently is.

**Comment by nirbheek** [comment]
> Having the same name of a library and an executable will fail if you build the project with msbuild on windows.

That will fail with any backend when you build with MSVC since pdb files use the `name` and not the full filename, so foo.dll and foo.exe write out foo.pdb and clobber each other. Right now we don't detect this case at all, partly because of limitations in ninja (it does not handle rules with multiple outputs well at all).

However, we shouldn't design according to the lowest-common-denominator, otherwise we'd have to assume every project is a flat layout on a case-insensitive layout built with MSVC, which is a setup with a lot of ridiculous problems. A lot of people do not care about MSVC at all, f.ex systemd, Xorg, etc, and would not be happy to be restricted by a platform they will never use.

One of the things we *could* do is warn (on all platforms) when some of these cases occur so people who are writing new libraries are aware of it and don't hit their head on it when they try to port their library to MSVC (for example). Of course, we should error out on the relevant platform.

Another thing we could do, is add a feature such as `supported_setups('platform=linux,osx,windows', 'layout=mirror,flat')`, and then you will get cross-platform errors about problems in your build files that are relevant to all these platforms or configurations. It will also tell users that they should expect to find problems when they try to build on an unsupported platform, so they can work with upstream to add that platform.

**Comment by sarum9in** [comment]
@nirbheek mentioned that related discussion is in #1867

TODO:
- Figure out corner cases besides Windows
- Make proof of concept

**Comment by nirbheek** [comment]
> Make proof of concept

fwiw, the linked PR also has an implementation.

**Comment by sarum9in** [comment]
Yup, but it is broken. I will study it anyway :) It's more or less note for myself because I tend to forget things when I switch contexts.

**Comment by sarum9in** [comment]
Discussed in IRC with @nirbheek and @jpakkane 

Things to consider:
- introspection ID must remain unique
- we start with simple name mangling `foo/baz/bar` becomes `foo_baz_bar`, we ignore `foo/baz_bar` vs `foo_baz/bar` for now
